### PR TITLE
Can multires

### PR DESCRIPTION
**Pull request by canergen**
Contains code for fixed multivariate analysis and fixed normalization of distances for the attention model.
 I have disabled the normalization for the linear model. Maybe we should have both.

### CODE REVIEW

File: src/scvi_v2/_module.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ class _DecoderZXAttention(nn.Module):
     n_out: int
     n_batch: int
     n_latent_sample: int = 16
-    h_activation: Callable = nn.softmax
+    h_activation: Callable = nn.softplus
     n_channels: int = 4
     n_heads: int = 2
```
Comment (PierreBoyeau) [inline src/scvi_v2/_module.py]: I switched back to softmax. are you using softplus in the local branch?

File: src/scvi_v2/_module.py [hunk 2]
```diff
@@ -584,3 +583,16 @@ def compute_h_from_x(self, x, sample_index, batch_index, cf_sample=None, continu
...
         return generative_outputs["h"]
+
+    def compute_h_from_x_eps(self, x, sample_index, batch_index, extra_eps, cf_sample=None, continuous_covs=None, mc_samples=10):
+        """Compute normalized gene expression from observations using predefined eps"""
+        library = 7.0 * jnp.ones_like(sample_index)  # placeholder, has no effect on the value of h.
+        inference_outputs = self.inference(x, sample_index, mc_samples=mc_samples, cf_sample=cf_sample, use_mean=False)
+        generative_inputs = {
+            "z": inference_outputs["z_base"] + extra_eps,
+            "library": library,
+            "batch_index": batch_index,
```
Comment (PierreBoyeau) [inline src/scvi_v2/_module.py:603]: we need to figure out why we need to start from `z_base` and not `u`. Do you have more insights on this?

File: src/scvi_v2/_model.py [hunk 3]
```diff
@@ -329,34 +331,41 @@ def per_sample_inference_fn(pair):
...
-                    normalization_vars = normalization_vars.reshape(-1, 1, 1)
-                    normalized_dists = np.clip(mean_dists - normalization_means, a_min=0, a_max=None) / (
-                        normalization_vars**0.5
-                    )
-
             if reqs.needs_sampled_distances:
                 sampled_dists = self._compute_distances_from_representations(sampled_zs_, indices, norm=norm)
+                
+            if reqs.needs_normalized_distances:
+                jit_inference_fn = self.module.get_jit_inference_fn(
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: nit: ideally we move the next 5 lines to its own function like before

File: src/scvi_v2/_model.py [hunk 4]
```diff
@@ -329,34 +331,41 @@ def per_sample_inference_fn(pair):
...
+                )
+                outputs = jit_inference_fn(self.module.rngs, array_dict)
+                sampled_zs_null = outputs["z"].transpose((1, 0, 2))
+                
+                dists_null = self._compute_distances_from_representations(
+                    sampled_zs_null, indices, norm=norm, return_xarray=False).reshape(len(indices), -1)
+                
+                # Please do experiment and check division by variance, not sure about its effect.
+                normalized_dists = ((sampled_dists - dists_null.mean(-1).reshape(-1, 1, 1, 1))/(
+                    dists_null.mean(-1).reshape(-1, 1, 1, 1))).mean('mc_sample')
```
Comment (justjhong) [inline src/scvi_v2/_model.py]: dividing by this seems to have the effect of making the minimum non-diagonal element pretty high (since dists_null.mean(-1) tends to be small). Can we try to clip it so that the minimum element is close to zero?

Comment (canergen) [reply src/scvi_v2/_model.py]: As described on Slack the issue is the number of mc_samples. Fix in next commit. I'm not sure about the percentile or mean.

File: src/scvi_v2/_module.py (hunk 2 above)
Comment (canergen) [reply src/scvi_v2/_module.py:603]: z_base is necessary as we allow the freedom that u and z_base have different dimensions. Using z is a problem because it contains two times epsilon in that case but z_base is the layer where the epsilons have an effect.

Comment (PierreBoyeau) [reply src/scvi_v2/_module.py:603]: Sorry for the confusing. This LGTM.

File: src/scvi_v2/_module.py [hunk 5]
```diff
@@ -348,6 +348,7 @@ class MrVAE(JaxBaseModuleClass):
     n_input: int
     n_sample: int
     n_batch: int
+    n_labels: int
     n_continuous_cov: int
```
Comment (PierreBoyeau) [inline src/scvi_v2/_module.py:352]: Would it be possible for you to include changes linked to supervised MrVI in another PR? This component seems more tricky than the rest!

**Comment by justjhong** [comment]
@cane11 @PierreBoyeau are the changes here all incorporated into main? i think pierre added the fix and I added the new normalization stuff

**Comment by PierreBoyeau** [comment]
> @cane11 @PierreBoyeau are the changes here all incorporated into main? i think pierre added the fix and I added the new normalization stuff

Hi Justin, all features excluding the supervised prior have been included into main.
In particular, main contains the updated DE functionalities for the supervised analyses.
